### PR TITLE
[Prometheus] Split up projects based on hosting mechanism.

### PR DESCRIPTION
**Pull request by Yun-Ting**
Fixes https://github.com/open-telemetry/opentelemetry-dotnet/issues/3057.

## Changes

The updated project setup is as follows:

- `OpenTelemetry.Exporter.Prometheus.Shared` contains the PrometheusExporter, PrometheusExporterOptions and serialization methods for all the shared classes to both AspNetCore and HttpListener projects. This project is *not* expected to be consumed directly by the end users. It serves as a clean way to group shared code for the below 2 projects for semantic reasons. As a result, I've deleted the Implementation folder that was in `OpenTelemetry.Exporter.Prometheus` project, and moved out the shared files which was previously in the folder to be immediate children of the project.

- `OpenTelemetry.Exporter.Prometheus.HttpListener` contains the PrometheusHttpListener and its options class. It aims to provide end users a quick setup experience for using OTel metrics with Prometheus Exporter locally. Please refer to [readme](https://github.com/open-telemetry/opentelemetry-dotnet/blob/304f27ba3adbec7eda63abb21682f14f6377b2ff/src/OpenTelemetry.Exporter.Prometheus.HttpListener/README.md) for more details.

- `OpenTelemetry.Exporter.Prometheus.AspNetCore` contains the middleware and its extensions for registering. This is for users trying to run PrometheusExporter in production environment. Refer to [readme](https://github.com/open-telemetry/opentelemetry-dotnet/blob/304f27ba3adbec7eda63abb21682f14f6377b2ff/src/OpenTelemetry.Exporter.Prometheus.AspNetCore/README.md) for details.

TODO:
Currently, OpenTelemetry.Exporter.Prometheus.AspNetCore.Tests, OpenTelemetry.Exporter.Prometheus.HttpListener.Tests, and OpenTelemetry.Exporter.Prometheus.Shared.Tests are green. I'll be working on fixing the CIs (not sure if I broke other tests with bad merge.)

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Exporter.Prometheus.AspNetCore/OpenTelemetry.Exporter.Prometheus.AspNetCore.csproj [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
+    <TargetFrameworks>netcoreapp3.1;net462</TargetFrameworks>
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus.AspNetCore/OpenTelemetry.Exporter.Prometheus.AspNetCore.csproj:5]: dont see a need for targetting net462 here, as this is specifically for asp.net core

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    <RunApiCompat>false</RunApiCompat>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry\OpenTelemetry.csproj" />
+    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Prometheus\OpenTelemetry.Exporter.Prometheus.csproj" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.1'">
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj:27]: is this needed for HttpListener?

File: src/OpenTelemetry.Exporter.Prometheus/README.md [hunk 3]
```diff
@@ -11,8 +11,18 @@
 
 ### Step 1: Install Package
 
+Depending on the hosting mechanism you will be using:
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus/README.md:14]: This is not clear..  We should document the usage under respective packages. (I left a qn earlier about the future of OpenTelemetry.Exporter.Prometheus package as well, that is related to this)

**Comment by Yun-Ting** [comment]
> Need to see description of the overall package architecture and dependencies, and the expected/proposed end user experience.

It depends on whether we think there will be use cases for extending directly on OpenTelemetry.Exporter.Prometheus project (not using the middleware or the listener.)
I'm currently leaning towards on not exposing the OpenTelemetry.Exporter.Prometheus itself as a nuget and in the OpenTelemetry.Exporter.Prometheus.AspNetCore, and OpenTelemetry.Exporter.Prometheus.HttpListener, referencing the shared files as links.
What is your take @cijothomas ?

File: test/OpenTelemetry.Exporter.Prometheus.Tests/OpenTelemetry.Exporter.Prometheus.Tests.csproj [hunk 4]
```diff
@@ -23,6 +23,8 @@
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.InMemory\OpenTelemetry.Exporter.InMemory.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Prometheus\OpenTelemetry.Exporter.Prometheus.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Extensions.Hosting\OpenTelemetry.Extensions.Hosting.csproj" />
+    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Prometheus.HttpListener\OpenTelemetry.Exporter.Prometheus.HttpListener.csproj" />
+    <ProjectReference Include="..\..\src\OpenTelemetry.Exporter.Prometheus.AspNetCore\OpenTelemetry.Exporter.Prometheus.AspNetCore.csproj" />
```
Comment (reyang) [inline test/OpenTelemetry.Exporter.Prometheus.Tests/OpenTelemetry.Exporter.Prometheus.Tests.csproj:27]:
```suggestion
    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Prometheus.AspNetCore\OpenTelemetry.Exporter.Prometheus.AspNetCore.csproj" />
```

File: docs/metrics/getting-started-prometheus-grafana/Program.cs [hunk 5]
```diff
@@ -31,9 +32,12 @@ public static void Main()
     {
         using var meterProvider = Sdk.CreateMeterProviderBuilder()
             .AddMeter("MyCompany.MyProduct.MyLibrary")
-            .AddPrometheusExporter(options => { options.StartHttpListener = true; })
+            .AddPrometheusExporter()
             .Build();
 
+        using var listener = new PrometheusHttpListener(meterProvider);
+        listener.Start();
+
```
Comment (reyang) [inline docs/metrics/getting-started-prometheus-grafana/Program.cs:40]: This seems a bit complicated than it has to be. Is it possible to make it work with just a single line (e.g. `meterProvider.AddPrometheusHttpListener()`)?

The object lifecycle management also seems complicated, e.g. what if the exporter got disposed before the listener?

**Comment by cijothomas** [comment]
> > Need to see description of the overall package architecture and dependencies, and the expected/proposed end user experience.
> 
> It depends on whether we think there will be use cases for extending directly on OpenTelemetry.Exporter.Prometheus project (not using the middleware or the listener.) I'm currently leaning towards on not exposing the OpenTelemetry.Exporter.Prometheus itself as a nuget and in the OpenTelemetry.Exporter.Prometheus.AspNetCore, and OpenTelemetry.Exporter.Prometheus.HttpListener, referencing the shared files as links. What is your take @cijothomas ?

https://github.com/open-telemetry/opentelemetry-dotnet/issues/3057 - the proposal from the issue is good.

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj [hunk 6]
```diff
@@ -0,0 +1,30 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
+    <TargetFrameworks>netcoreapp3.1;net462</TargetFrameworks>
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj:5]: Might be able to be `netstandard2.0;net462` or _maybe_ even just `netstandard2.0`?

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs [hunk 7]
```diff
@@ -19,36 +19,45 @@
...
 {
-    /// <summary>
-    /// An HTTP listener used to expose Prometheus metrics.
-    /// </summary>
-    internal sealed class PrometheusExporterHttpServer : IDisposable
+    public sealed class PrometheusHttpListener : IDisposable
     {
         private readonly PrometheusExporter exporter;
-        private readonly HttpListener httpListener = new();
+        private readonly System.Net.HttpListener httpListener = new();
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs:29]:
```suggestion
        private readonly HttpListener httpListener = new();
```

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs [hunk 8]
```diff
@@ -59,16 +68,16 @@ public PrometheusExporterHttpServer(PrometheusExporter exporter)
...
-            foreach (string prefix in exporter.Options.HttpListenerPrefixes)
+            foreach (string prefix in prometheusHttpListenerOptions.HttpListenerPrefixes)
             {
                 this.httpListener.Prefixes.Add($"{prefix.TrimEnd('/')}{path}");
             }
         }
 
         /// <summary>
-        /// Start Http Server.
+        /// Start the http Listener.
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs:78]:
```suggestion
        /// Start the HttpListener.
```

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs [hunk 9]
```diff
@@ -88,7 +97,7 @@ public void Start(CancellationToken token = default)
         }
 
         /// <summary>
-        /// Stop exporter.
+        /// Stop the http listener.
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs:100]:
```suggestion
        /// Stop the HttpListener.
```

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs [hunk 10]
```diff
@@ -105,13 +114,23 @@ public void Stop()
...
+        {
+            if (!this.disposed)
             {
-                this.Stop();
-                this.httpListener.Close();
+                if (disposing && this.httpListener != null && this.httpListener.IsListening)
+                {
+                    this.Stop();
+                    this.httpListener.Close();
+                    this?.Dispose();
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs:130]: Is the null conditional necessary here?

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj (hunk 6 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj]: yes, I'm guessing using netstandard2.0 should be sufficient here.

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs (hunk 10 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs]: Nope, I've removed the null check for this.Dispose(). Thanks!

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs (hunk 7 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs:28]: Will get this waring without using System.Net for it :(.
What are the suggested ways to not spelling out the entire namespace for System.Net.HttpListener while not changing the PrometheusHttpListener namespace?
image

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj (hunk 6 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj]: I stumbled across the Castle.Core dependency issue while using netstandard2.0. Will do more digging and report back if I can fix it.

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj (hunk 2 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj]: I've cleaned it up :-)

File: src/OpenTelemetry.Exporter.Prometheus.AspNetCore/OpenTelemetry.Exporter.Prometheus.AspNetCore.csproj (hunk 1 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.Prometheus.AspNetCore/OpenTelemetry.Exporter.Prometheus.AspNetCore.csproj]: fixed

File: docs/metrics/getting-started-prometheus-grafana/Program.cs (hunk 5 above)
Comment (Yun-Ting) [reply docs/metrics/getting-started-prometheus-grafana/Program.cs]: Updated based on the suggestion; please check [it](https://github.com/open-telemetry/opentelemetry-dotnet/blob/be87d4478b016e749bc7641e28db6c18756815fa/src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusExporterHttpListenerMeterProviderBuilderExtensions.cs) out.

**Comment by Yun-Ting** [comment]
> @Yun-Ting Could you update the description with the final list of packages and their connections? It is not clear what would happend to the package OpenTelemetry.Exporter.Prometheus... Do we continue to publish it? Who is it meant for? Can one use it directly or its not meant for end user consumption?

@cijothomas , I've updated the PR description to answer these questions. I don't think we'll need to publish `OpenTelemetry.Exporter.Prometheus.Shared` as a nuget package. 
We will have 2 separate nuget packages, which are `OpenTelemetry.Exporter.Prometheus.HttpListener` and `OpenTelemetry.Exporter.Prometheus.AspNetCore`.

File: docs/metrics/getting-started-prometheus-grafana/getting-started-prometheus-grafana.csproj [hunk 11]
```diff
@@ -1,5 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <ItemGroup>
-    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Prometheus\OpenTelemetry.Exporter.Prometheus.csproj" />
+    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Prometheus.Shared\OpenTelemetry.Exporter.Prometheus.Shared.csproj" />
```
Comment (cijothomas) [inline docs/metrics/getting-started-prometheus-grafana/getting-started-prometheus-grafana.csproj:3]: This reference to the Shared project is incorrect, as it is not meant to be publicly available.

File: src/OpenTelemetry.Exporter.Prometheus.AspNetCore/README.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    });
+    ```
+
+* Or configure directly:
+
+    Call the `AddPrometheusExporter` `MeterProviderBuilder` extension to
+    register the Prometheus exporter.
+
+    ```csharp
+    using var meterProvider = Sdk.CreateMeterProviderBuilder()
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus.AspNetCore/README.md:34]: would be nicer to show meterProvider being added to DI manually here. Else users who copy this as-is, will let the provider gets GCed!

File: src/OpenTelemetry.Exporter.Prometheus.AspNetCore/README.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+The `PrometheusExporter` can be configured using the `PrometheusExporterOptions`
+properties. Refer to
+[`TestPrometheusExporter.cs`](../../examples/Console/TestPrometheusExporter.cs)
+for example use.
+
+### ScrapeEndpointPath
+
+Defines the path for the Prometheus scrape endpoint for
+either the HTTP listener or the middleware registered by
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus.AspNetCore/README.md:84]: the comment seems outdated, as it is referring to Http Listener?

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListenerOptions.cs [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+{
+    public class PrometheusHttpListenerOptions
+    {
+        private IReadOnlyCollection<string> httpListenerPrefixes = new string[] { "http://localhost:9464/" };
+
+        /// <summary>
+        /// Gets or sets the prefixes to use for the http listener. Default
+        /// value: http://localhost:9464/.
+        /// </summary>
+        public IReadOnlyCollection<string> HttpListenerPrefixes
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListenerOptions.cs:31]: This'd be apt time to rename this to simply "Prefixes", as the class name is already qualifying this.

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusExporterHttpListenerMeterProviderBuilderExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+using OpenTelemetry.Exporter.Prometheus.HttpListener;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Metrics
+{
+    public static class PrometheusExporterHttpListenerMeterProviderBuilderExtensions
+    {
+        public static MeterProviderBuilder AddPrometheusHttpListener(
+            this MeterProviderBuilder builder,
+            Action<PrometheusExporterOptions> configureExporterOptions = null,
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusExporterHttpListenerMeterProviderBuilderExtensions.cs:28]: Where is `PrometheusExporterOptions` class defined? Why we need two separate options class?

**Review by cijothomas — CHANGES_REQUESTED** [review]
See comments. Very close, but some refinements needed in the shared class usage, and public API

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListenerOptions.cs (hunk 14 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListenerOptions.cs:31]: Resolved in https://github.com/open-telemetry/opentelemetry-dotnet/pull/3430.

**Comment by Yun-Ting** [comment]
Closing this PR - please review the new https://github.com/open-telemetry/opentelemetry-dotnet/pull/3430 instead. Thanks!

File: src/OpenTelemetry.Exporter.Prometheus.AspNetCore/README.md (hunk 13 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.Prometheus.AspNetCore/README.md:84]: Resolved in https://github.com/open-telemetry/opentelemetry-dotnet/pull/3430.

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj (hunk 6 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.Prometheus.HttpListener/OpenTelemetry.Exporter.Prometheus.HttpListener.csproj]: Updated to be using netstandard2.0;net462 here: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3430

**Pull request closed without merging**